src/rules/objectLiteralSortKeysRule.ts

```diff
@@ -15,74 +15,170 @@
+    public static SHORTHAND_FAILURE_FACTORY(name: string) {
+        return `Shorthand property '${name}' should come before normal properties`;
+    }
+
+    public static ALHPA_FAILURE_FACTORY(name: string) {
         return `The key '${name}' is not sorted alphabetically`;
     }
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithFunction(sourceFile, walk, {
-            ignoreCase: this.ruleArguments.indexOf(OPTION_IGNORE_CASE) !== -1,
-        });
+        return this.applyWithFunction(sourceFile, walk, parseOptions(this.ruleArguments));
     }
 }
 
-function walk(ctx: Lint.WalkContext<Options>) {
-    return ts.forEachChild(ctx.sourceFile, function cb(node): void {
-        if (isObjectLiteralExpression(node) && node.properties.length > 1 &&
-            !isSameLine(ctx.sourceFile, node.properties.pos, node.end)) {
-            let lastKey: string | undefined;
-            const {options: {ignoreCase}} = ctx;
-            outer: for (const property of node.properties) {
-                switch (property.kind) {
-                    case ts.SyntaxKind.SpreadAssignment:
-                        lastKey = undefined; // reset at spread
+function parseOptions(ruleArgs: any[]): IOptions {
+    const options: IOptions = {
+        checkSingleLine: ruleArgs.indexOf(OPTION_CHECK_SINGLE_LINE) !== -1,
+        ignoreCase: ruleArgs.indexOf(OPTION_IGNORE_CASE) !== -1,
+        shorthandFirst: ruleArgs.indexOf(OPTION_SHORTHAND_FIRST) !== -1,
+    };
+    return options;
+}
+
+function walk(ctx: Lint.WalkContext<IOptions>) {
+    const { options, sourceFile } = ctx;
+    const { checkSingleLine, ignoreCase, shorthandFirst } = options;
+
+    function canCheckNode(node: ts.Node) {
+        // Only check object literals with at least one key
+        if (!isObjectLiteralExpression(node) || node.properties.length <= 1) {
+            return false;
+        }
+
+        // Only check single-line object literals if explicitly asked
+        if (!checkSingleLine && isSameLine(sourceFile, node.properties.pos, node.end)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    function shouldCheckSyntax(currentProperty: SortableProperty, lastProperty?: SortableProperty) {
+        // Only check syntax if option is set, we have a property to check ,and they changed
+        if (!shorthandFirst || lastProperty === undefined) {
+            return false;
+        }
+        return currentProperty.kind !== lastProperty.kind;
+    }
+
+    function cb(node: ts.Node): void {
+        if (!canCheckNode(node)) {
+            return ts.forEachChild(node, cb);
+        }
+
+        let lastKey: string | undefined;  // to look back for spelling
+        let lastProperty: SortableProperty | undefined; // to look back for assignment
+
+        const { properties } = node as ts.ObjectLiteralExpression;
+        outer: for (const property of properties) {
```

Comment:
😱 is there any way to avoid this `outer:` syntax? perhaps another helper function?